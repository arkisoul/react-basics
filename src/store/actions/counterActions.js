export const INCREMENT = "INCREMENT";
export const DECREMENT = "DECREMENT";
export const INCREMENTBYVALUE = "INCREMENTBYVALUE";

// Action creators are functions that returns an action object
export const incrementCounter = () => ({ type: INCREMENT });
export const decrementCounter = () => ({ type: DECREMENT });
export const incrementCounterByValue = (value) => ({
  type: INCREMENTBYVALUE,
  payload: { value },
});
