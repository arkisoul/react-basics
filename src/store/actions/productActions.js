import axios from 'axios';

export const ADDPRODUCT = "ADDPRODUCT";
export const REMOVEPRODUCT = "REMOVEPRODUCT";
export const PRODUCT_REQUEST_START = "PRODUCT_REQUEST_START";
export const PRODUCT_REQUEST_SUCCESS = "PRODUCT_REQUEST_SUCCESS";
export const PRODUCT_REQUEST_FAILURE = "PRODUCT_REQUEST_FAILURE";

export const addProduct = (product) => ({
  type: ADDPRODUCT,
  payload: { product: product },
});
export const removeProduct = (productId) => ({
  type: REMOVEPRODUCT,
  payload: { productId: productId },
});
export const productFetchStart = () => ({ type: PRODUCT_REQUEST_START });
export const productFetchSuccess = (products) => ({
  type: PRODUCT_REQUEST_SUCCESS,
  payload: { products },
});
export const productFetchFailure = (error) => ({
  type: PRODUCT_REQUEST_FAILURE,
  payload: { error },
});

export const fetchProducts = () => async (dispatch) => {
  try {
    dispatch(productFetchStart());
    const res = await axios.get("https://fakestoreapi.com/products");
    const products = res.data;
    dispatch(productFetchSuccess(products))
  } catch (error) {
    dispatch(productFetchFailure(error))
  }
};
