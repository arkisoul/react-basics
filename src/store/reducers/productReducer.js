import * as actions from "../actions/productActions";

export const initialState = {
  products: [],
  loading: false,
  error: null,
};

export const productReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case actions.ADDPRODUCT:
      return { ...state, products: [...payload.product] };

    case actions.REMOVEPRODUCT:
      const filteredProducts = [...state.products].filter(
        (product) => product.id !== payload.productId
      );
      return { ...state, products: filteredProducts };

    case actions.PRODUCT_REQUEST_START:
      return { ...state, loading: true };

    case actions.PRODUCT_REQUEST_SUCCESS:
      return { ...state, loading: false, products: payload.products };

    case actions.PRODUCT_REQUEST_FAILURE:
      return { ...state, loading: false, error: payload.error };

    default:
      return state;
  }
};
