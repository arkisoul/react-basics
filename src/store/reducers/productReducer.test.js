import { describe, it, expect } from "@jest/globals";

import { initialState, productReducer } from "./productReducer";
import * as productActions from "../actions/productActions";

describe("product reducer", () => {
  const product = {
    id: 1,
    title: "Product title",
    body: "Product description",
    price: 12.99,
  };

  it("should return initial state on store creation", () => {
    expect(productReducer(undefined, {})).toEqual(initialState);
    expect(productReducer(undefined, {})).toMatchSnapshot();
  });

  it("should return updated state on add product action", () => {
    expect(
      productReducer(undefined, productActions.addProduct([product]))
    ).toEqual({
      products: [product],
      loading: false,
      error: null,
    });
    expect(
      productReducer(undefined, productActions.addProduct([product]))
    ).toMatchSnapshot();
    expect(
      productReducer(
        {
          loading: false,
          error: null,
          products: [product],
        },
        productActions.addProduct([
          {
            id: 2,
            title: "another product",
            body: "another product description",
            price: 9.99,
          },
        ])
      )
    ).toEqual({
      products: [
        {
          id: 2,
          title: "another product",
          body: "another product description",
          price: 9.99,
        },
      ],
      loading: false,
      error: null,
    });
    expect(
      productReducer(
        {
          loading: false,
          error: null,
          products: [product],
        },
        productActions.addProduct([
          {
            id: 2,
            title: "another product",
            body: "another product description",
            price: 9.99,
          },
        ])
      )
    ).toMatchSnapshot();
  });

  it("should return updated state on remove product action", () => {
    expect(
      productReducer({
      products: [product],
      loading: false,
      error: null,
    }, productActions.removeProduct(product.id))).toEqual({
      products: [],
      loading: false,
      error: null,
    });
    expect(
      productReducer(
        {
          products: [product],
          loading: false,
          error: null,
        },
        productActions.removeProduct(2)
      )
    ).toEqual({
      products: [product],
      loading: false,
      error: null,
    });
  });
});
