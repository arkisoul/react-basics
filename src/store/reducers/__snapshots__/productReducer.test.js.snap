// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product reducer should return initial state on store creation 1`] = `
Object {
  "error": null,
  "loading": false,
  "products": Array [],
}
`;

exports[`product reducer should return updated state on add product action 1`] = `
Object {
  "error": null,
  "loading": false,
  "products": Array [
    Object {
      "body": "Product description",
      "id": 1,
      "price": 12.99,
      "title": "Product title",
    },
  ],
}
`;

exports[`product reducer should return updated state on add product action 2`] = `
Object {
  "error": null,
  "loading": false,
  "products": Array [
    Object {
      "body": "another product description",
      "id": 2,
      "price": 9.99,
      "title": "another product",
    },
  ],
}
`;
